File: Sources/SPFNUI/Components/HeaderIcons.swift
#if canImport(SwiftUI)
// SPFN Mobile — the two marks the SDK still draws in the navigation bar.
//
// Counterpart of
// android/spfn-ui/src/main/kotlin/xyz/superfunction/spfn/ui/components/HeaderIcons.kt. The
// back on iOS is the system navigation bar's own button (`Screen.swift`), so neither mark is
// part of an ordinary screen any more. Each has one caller left:
//
//   - the X is the flow's close on the root of a presented flow, the SDK's item in the bar's
//     trailing place — an X on the right is what a person on an iPhone looking for the way
//     out of a sheet looks for (decision N3);
//   - the chevron is the back on the root of a pushed flow drawn WITHOUT a `NavigationHost`,
//     the compatibility path `FlowHost.swift` describes. That root is its own navigator's
//     root, so the system draws no back button there, and without this mark a person would
//     be left on a screen with no way off it (docs/IMPLEMENTATION-PITFALLS.md P31).
//
// Not part of the public component set, and deliberately outside the names section 15 of the
// validator compares. These are marks the SDK draws for itself; a host app that wants its own
// control passes an item to `Screen`, which is the door that already exists.
//
// The marks are 20pt. The system bar sizes the button around them, and the bar's own buttons
// meet the platform's minimum touch target, `Metrics.touchTarget`, which is P21's number.
//
// iOS draws both out of SF Symbols, which is the system's own vocabulary for exactly these
// two marks; the Compose half has no such vocabulary that is not Material (decision C2), so it
// draws the same two shapes with `Canvas`. What the two halves share is the geometry — 20
// across, and the primary text colour.

import SwiftUI

/// The mark a back control draws.
struct BackChevron: View
{
    @Environment(\.spfnTheme) private var theme
    @Environment(\.colorScheme) private var scheme

    var body: some View
    {
        Image(systemName: "chevron.left")
            .font(.system(size: Metrics.iconSize, weight: .medium))
            .foregroundStyle(theme.palette(for: scheme).text)
    }
}

/// The mark a close control draws.
struct CloseCross: View
{
    @Environment(\.spfnTheme) private var theme
    @Environment(\.colorScheme) private var scheme

    var body: some View
    {
        Image(systemName: "xmark")
            .font(.system(size: Metrics.iconSize, weight: .medium))
            .foregroundStyle(theme.palette(for: scheme).text)
    }
}
#endif

File: android/spfn-ui/src/main/kotlin/xyz/superfunction/spfn/ui/components/HeaderIcons.kt
// SPFN Mobile — the two marks a header draws for the two ways out.
//
// Counterpart of Sources/SPFNUI/Components/HeaderIcons.swift. A chevron on the left for a
// back, an X on the right for a close (decision N3): the words `Back` and `Close` in body
// type read as prose rather than as controls, and a person looking for the way out of a
// sheet looks at the top right corner.
//
// Not part of the public component set, and deliberately outside the nine names section 15
// of the validator compares. These are two marks this component draws for itself; a host app
// that wants its own control passes a slot to `Screen`, which is the door that already
// exists.
//
// `Canvas` and not Material. This repository depends on no Material artifact (decision C2),
// so `Icons.Default.Close` is not a thing this module may reach for, and two strokes are
// two strokes: what has to match the iOS half is the GEOMETRY — 20dp across, inside the
// platform's minimum touch target, in the primary text colour — and not which glyph table
// they came out of.
//
// The size split is P21's, stated once here and once on the SwiftUI half: the MARK is 20dp
// and the frame around it is `Metrics.TOUCH_TARGET`. A control drawn at the mark's own size
// reports a rectangle its neighbour has already eaten, and a device runner then taps the
// neighbour.

package xyz.superfunction.spfn.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.unit.dp

/** The mark a back control draws: one chevron, pointing the way it goes. */
@Composable
internal fun BackChevron()
{
    val colour = spfnPalette().text;
    Canvas(modifier = Modifier.size(Metrics.ICON_SIZE))
    {
        // Inset so the stroke's own width stays inside the 20dp box: a line drawn to the
        // edge is half a line outside it, and the two marks would then sit at two
        // different visual sizes.
        val inset = size.minDimension * CHEVRON_INSET;
        val middle = size.height / 2f;
        drawLine(
            color = colour,
            start = Offset(size.width - inset, inset),
            end = Offset(inset, middle),
            strokeWidth = STROKE.toPx(),
            cap = StrokeCap.Round
        );
        drawLine(
            color = colour,
            start = Offset(inset, middle),
            end = Offset(size.width - inset, size.height - inset),
            strokeWidth = STROKE.toPx(),
            cap = StrokeCap.Round
        );
    }
}

/** The mark a close control draws: two strokes across the same box. */
@Composable
internal fun CloseCross()
{
    val colour = spfnPalette().text;
    Canvas(modifier = Modifier.size(Metrics.ICON_SIZE))
    {
        val inset = size.minDimension * CROSS_INSET;
        drawLine(
            color = colour,
            start = Offset(inset, inset),
            end = Offset(size.width - inset, size.height - inset),
            strokeWidth = STROKE.toPx(),
            cap = StrokeCap.Round
        );
        drawLine(
            color = colour,
            start = Offset(size.width - inset, inset),
            end = Offset(inset, size.height - inset),
            strokeWidth = STROKE.toPx(),
            cap = StrokeCap.Round
        );
    }
}

/** How thick both marks are drawn. Not a token: it is these two shapes' own arithmetic. */
private val STROKE = 2.dp;

/** How far the chevron's ends stand inside its box, as a fraction of it. */
private const val CHEVRON_INSET: Float = 0.28f;

/** The same for the cross, which reaches further because it has two ends to spare. */
private const val CROSS_INSET: Float = 0.22f;
